airflow/models/pool.py

```diff
@@ -48,6 +50,45 @@ def get_pool(pool_name, session=None):
     def get_default_pool(session=None):
         return Pool.get_pool(Pool.DEFAULT_POOL_NAME, session=session)
 
+    @staticmethod
+    @provide_session
+    def slots_stats(session=None) -> Dict[str, Dict[str, int]]:
+        from airflow.models.taskinstance import TaskInstance  # Avoid circular import
+
+        pools = {}
+
+        pool_rows = session.query(Pool.pool, Pool.slots).all()
+        for (pool_name, total_slots) in pool_rows:
+            pools[pool_name] = {
+                "total": total_slots,
+                "used": 0,
+                "queued": 0,
+            }
+
+        state_count = (
+            session.query(TaskInstance.pool, TaskInstance.state, func.count())
+            .filter(TaskInstance.state.in_(STATES_TO_COUNT_AS_RUNNING))
+            .filter(TaskInstance.pool.in_(pools.keys()))
```

Comment:
Is this conditional needed?

Reply:
Yes, there is a pool can be deleted before task state is updated. However, I think we can handle this case in a better way by doing the validation in python in the subsequent for loop.